src/Interpreters/Set.cpp

```diff
@@ -331,13 +403,39 @@ ColumnPtr Set::execute(const ColumnsWithTypeAndName & columns, bool negative) co
             result = castColumnAccurate(column_to_cast, data_types[i], cast_cache.get());
         }
 
-        materialized_columns.emplace_back() = result;
-        key_columns.emplace_back() = materialized_columns.back().get();
+        // If the original column is DateTime64, check for sub-second precision
+        if (isDateTime64(column_to_cast.column->getDataType()))
+        {
+            ColumnPtr filtered_null_map_column = checkDateTimePrecision(column_to_cast);
+
+            // Extract existing null map and nested column from the result
+            const ColumnNullable * result_nullable_column = typeid_cast<const ColumnNullable *>(result.get());
+            const IColumn * nested_result_column = result_nullable_column
+                ? &result_nullable_column->getNestedColumn()
+                : result.get();
+
+            ColumnPtr existing_null_map_column = result_nullable_column
+                ? result_nullable_column->getNullMapColumnPtr()
+                : nullptr;
+
+            ColumnPtr merged_null_map_column = mergeNullMaps(existing_null_map_column, filtered_null_map_column);
```

Comment:
I think that in case of `transform_null_in` we
1. Should not merge masks
2. should not extractNestedColumnsAndNullMap